app/services/chart_builder.rb

```diff
@@ -1,15 +1,36 @@
 class ChartBuilder
-  def initialize(benchmark_runs)
+  attr_reader :data, :versions, :benchmark_result_type
+
+  def initialize(benchmark_runs = [], benchmark_result_type)
+    @benchmark_result_type = benchmark_result_type
     @benchmark_runs = benchmark_runs
+    # @versions is a hash that holds all version/environment information for each datapoint, 
+    #   and gets converted to html and stored in @data[:categories]
+    @versions = []
+    # @data[:columns] is JSON that looks like [{ name: "benchmark1", data: [1.1, 1.2] }]
+    # @data[:categories] is JSON that looks like [version_html_string, version_html_string]
     @data = {}
     @data[:columns] = {}
   end
 
+  # `cache_read` looks like 
+  # { datasets: [{ name: "benchmark1", data: [1.1, 1.2] }], versions: [version_hash, version_hash] }
+  def construct_from_cache(cache_read)
+    @versions = cache_read[:versions]
+    @data[:categories] = @versions.map { |version| hash_to_html(version) }.to_json
+    @data[:columns] = cache_read[:datasets].to_json
+    self
+  end
+
   def build_columns
     @benchmark_runs.each do |benchmark_run|
       if block_given?
+        version = yield(benchmark_run)
         @data[:categories] ||= []
-        @data[:categories] << yield(benchmark_run)
+        if version != @versions.last
```

Comment:
if there's two datalines on a chart, ex. [1,2,3] and [4,5,6], then those get processed here in the order 1, 4, 2, 5, 3, 6, since we sort by versions. 1 and 4 have the same version - we don't want repeats in `@versions`